module/state_synchronization/adder.go

```diff
@@ -0,0 +1,285 @@
+package state_synchronization
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/ipfs/go-cid"
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/model/encoding"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module"
+	"github.com/onflow/flow-go/module/blobs"
+	"github.com/onflow/flow-go/network"
+)
+
+const defaultBlobBatchSize = 16
+
+type ExecutionDataAdderOption func(*executionDataAdderImpl)
+
+func WithBlobSizeLimit(size int) ExecutionDataAdderOption {
+	return func(s *executionDataAdderImpl) {
+		s.maxBlobSize = size
+	}
+}
+
+func WithBlobBatchSize(batchSize int) ExecutionDataAdderOption {
+	return func(s *executionDataAdderImpl) {
+		s.blobBatchSize = batchSize
+	}
+}
+
+func NewExecutionDataAdder(
+	codec encoding.Codec,
+	compressor network.Compressor,
+	blobService network.BlobService,
+	logger zerolog.Logger,
+	metrics module.ExecutionDataAdderMetrics,
+	opts ...ExecutionDataAdderOption,
+) *executionDataAdderImpl {
+	eda := &executionDataAdderImpl{
+		logger.With().Str("component", "execution_data_adder").Logger(),
+		metrics,
+		DefaultMaxBlobSize,
+		defaultBlobBatchSize,
+		&serializer{codec, compressor},
+		blobService,
+	}
+
+	for _, opt := range opts {
+		opt(eda)
+	}
+
+	return eda
+}
+
+var _ ExecutionDataAdder = (*executionDataAdderImpl)(nil)
+
+type executionDataAdderImpl struct {
+	logger        zerolog.Logger
+	metrics       module.ExecutionDataAdderMetrics
+	maxBlobSize   int
+	blobBatchSize int
+	serializer    *serializer
+	blobService   network.BlobService
+}
+
+func (s *executionDataAdderImpl) AddExecutionDataRoot(
```

Comment:
I'm proposing a refactor of the `Getter` and `Adder` implementation. This is the pseudo code.

The main benefit of the refactor is to provide high conhesion abstractions, so that the function body of `AddExecutionDataRoot` and `GetChunkExecutionDatas` is simplified and self-explains the main logic. 
```golang
func Chunkify(ComputationResult) []*ChunkExectuionData

type ChunkifiedExecutionData struct {
  BlockID flow.Identifier
  Chunks []*ChunkExecutionData
}

type ExecutionDataRoot struct {
  BlockID flow.Identifier
  ChunkCIDs []cid.CID
}

type ChunkExecutionData struct { 	Collection *flow.Collection
	Events     flow.EventsList
	TrieUpdate *ledger.TrieUpdate
}

// AddExecutionData adds the given list of chunk execution data to blob service, and return the root CID of 
// chunksExecutionData is generated from execution.ChunkifyExecutionData(computationResult), which should be a stateless computation
//  from some other package. 
func (adder *ExecutionDataAdder) AddChunkifiedExecutionData(ctx, executionData *ChunkifiedExecutionData) (flow.Identifier, error) {
  // adding ChunkExectuionData could be possibly run concurrently, splitting the concurrency control and 
  // a single job operation into, and using by composing them. 
  chunkCIDs := runConcurrently(executionData.Chunks, func(chunkExecutionData) (cid.CID, error) {
    return adder.addChunkExecutionDataToBlobService(ctx, chunkExecutionData)
  })
  
  // the reason we encode the block ID and chunk CIDs together is so that the blockID could be used
  // by the Getter to verify the rootCID.   rootBlob := adder.codec.Encode(&ExecutionDataRoot{     BlockID: executionData.Chunks,
    ChunkCIDs: chunkCIDs,
  })
  // compute the root CID of all the execution data  chunks
  rootCIDIdentifier := flow.CIDToID(rootBlob.CID())   return rootCIDIdentifier, nil } 
// addChunkExecutionData adds a single chunkExectuionData to the blobService, and return the Cid 
// for the given chunkExecutionData func (*ExecutionDataAdder) addChunkExecutionDataToBlobService(ctx, chunkExecutionData) cid.CID, error

// symmetric Encode and Decode methods to convert between Blob and ExecutionDataRoot
type ExecutionDataRootCodec interface {
	Encode(*ExecutionDataRootCodec) blobs.Blob
       Decode(blob *blobs.Blob) *ExecutionDataRootCodec
}

// GetChunkExecutionDatas gets all the chunk execution data for the given rootCID. The caller must ensure
// the validity of blockID and rootCIDIdentifier, which is supposed to be taken from sealed result to guarantee
// validity. 
// The blockID is taken for sanity check. If `MismatchedBlockIDError` error is returned, the caller should treat it
// as a fatal error. 
func (*Getter) GetChunkExecutionDatas(ctx, blockID, rootCIDIdentifier) ([]*ChunkExecutionData, error) {
  // rootID is a bit confusing as if it was the root hash of the trie.
  // rootCIDIdentifier although is a bit longer but more clear about what it actually is.
  rootCID := flow.IDtoCID(rootCIDIdentifier)

  // an invalid rootCID would fail to retrieve its blob
  blob, err := getter.blobGetter.GetBlob(ctx, rootCID)
  executionDataRoot := getter.codec.Decode(blob)

  // is this check not strictly necessary, but as a sanity check, because the caller is supposed to call with a good (blockID, rootCID) 
  // from a sealed result. Fail this check should be a fatal error, and the caller is responsible to handle this sentinel error properly.
  if blockID != executionDataRoot.BlockID {     return nil, MismatchedBlockIDError
  }
  
  chunks, err := runConcurrently(executionDataRoot.ChunkCIDs, func(cid) (*ChunkExecutionData, error) {
    return getter.getChunkExecutionDataByCID(ctx, cid)
  })

  return chunks, err
}

func (*Getter) getChunkExecutionDataByCID(ctx, cid) (*ChunkExecutionData, error)
```